### PR TITLE
Task 4 http servers

### PR DESCRIPTION
**Pull request by Reshetnyak**

### CODE REVIEW

File: app/server/helpers/isObject.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export function isObject(obj) {
+    return Object.prototype.toString.call(obj) === '[object Object]';
```
Comment (vdyalex) [inline app/server/helpers/isObject.js:2]: Consider using the comparison with `typeof obj === 'object'`.

File: app/server/http-servers/echo-server.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+const port = 8084;
+
+http.createServer((req, res) => {
+  res.writeHead(200);
+  req.pipe(res);
+})
+.listen(port, errorHandler);
+
+function errorHandler(err) {
+    if (err) {
```
Comment (vdyalex) [inline app/server/http-servers/echo-server.js]: Can be also:

```javascript
return err
  ? console.log(err)
  : console.log(`Server is listening on port ${port}`);
```

File: app/server/http-servers/html-server.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+const noIndexTemplate = `<html><head></head><body><h1>Somebody has stolen our template!!!</h1></body></html>\n`;
+
+http.createServer( requestHandler )
+    .listen(port, errorHandler);
+
+function requestHandler(req, res) {
+
+    const template = getTemplate('index.html');
+
+    res.writeHead(200, {'Content-Type': 'text/html'});
```
Comment (vdyalex) [inline app/server/http-servers/html-server.js]: If the template doesn't exist, would make sense return an appropriate HTTP status code. It isn't required, but can be considered as a part of best practices in API development :)

File: app/server/http-servers/html-server.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    if (template) {
+        _streamFromString(template)
+            .pipe( through( (chunk, enc, cb) => {
+                cb(null, setMessage(chunk.toString(), 'I wish you all the best!'));
```
Comment (vdyalex) [inline app/server/http-servers/html-server.js]: Avoid callback hell and spaghetti code:

```javascript
const stream = through((chunk, _, callback) => {
  const message = setMessage(chunk.toString(), 'I wish you all the best!')
  callback(null, message);
});

_streamFromString(template)
  .pipe(stream)
  .pipe(res);
```

**Review by vdyalex — APPROVED** [review]
I've checked the task. Your implementation is good. The task created by CDP is completely non-sense, mixing responsibilities, like create an `express` server in parallel with `http` module implementations, mocking a kind of unexisting resources only to meet lecture's topics... It shall be discussed with CDP after the mentoring completion. For the moment, it's OK with your implementation.

File: app/server/helpers/isObject.js (hunk 1 above)
Comment (Reshetnyak) [reply app/server/helpers/isObject.js:2]: I'm using it for checking is passed object is actually an object. If `obj = []`, `typeof obj === 'object'` would also be true. It doesn't fit for my particular task.

Comment (vdyalex) [reply app/server/helpers/isObject.js:2]: Good point! There is also another way to simplify that logic (not this function in particular, but the way to build the flow).

You can export the modules as functions, when each method receives the argument of `app`, like:
```javascript
module.export = (app) => {
  app.use(...)
}
```

...and  in the index file:

```javascript
const express = require('express');
const app = express();
const myRoute = require('./my-route');

myRoute(app);
```

In this way, the exporting would not be required, avoiding mix the routes with middlewares and eliminating the necessity to test if it's an object or not. However, this is a solution that would work for a reasonable task. I don't recommend you to implement it here, although you may try.

**Pull request merged**